packages/actor-optimize-link-traversal-filter-tree-links/lib/ActorOptimizeLinkTraversalFilterTreeLinks.ts

```diff
@@ -0,0 +1,203 @@
+    return filterExist && relations.length > 0 ?
+      Promise.resolve(true) :
+      Promise.reject(new Error(
+        'the action must contain TREE relation and the query must contain at least a filter',
+      ));
+  }
+
+  public async run(action: IActionOptimizeLinkTraversal): Promise<IActorOptimizeLinkTraversalOutput> {
+    const filterMap: Map<string, boolean> = new Map();
+    // Extract the filter expression
+    const filterOperation: Algebra.Expression = (() => {
+      const query: Algebra.Operation = action.context.get(KeysInitQuery.query)!;
+      return query.input.expression;
+    })();
+
+    // Extract the bgp of the query
+    const queryBody: RDF.Quad[] = this.findBgp(action.context.get(KeysInitQuery.query)!);
+    if (queryBody.length === 0) {
+      return { filters: filterMap };
+    }
+    // Capture the relation from the input
+    const relations: IRelation[] = (() => {
+      if (typeof action.treeMetadata !== 'undefined') {
+        return action.treeMetadata.relation!;
+      }
+      return [];
+    })();
+    for (const relation of relations) {
+      if (typeof relation.path === 'undefined' || typeof relation.value === 'undefined') {
+        filterMap.set(relation.node, true);
+        continue;
+      }
+      // Find the quad from the bgp that are related to the TREE relation
+      const relevantQuads = this.findRelavantQuad(queryBody, relation.path.value);
+      if (relevantQuads.length === 0) {
+        filterMap.set(relation.node, true);
+        continue;
+      }
+
+      // Create the binding in relation to the relevant quad
+      const bindings = this.createBinding(relevantQuads, relation.value.quad);
+      const filterExpression: Algebra.Operation = this.deleteUnrelevantFilter(filterOperation, bindings);
+      if (filterExpression.args.length === 0) {
+        filterMap.set(relation.node, true);
+        continue;
+      }
+      const evaluator = new AsyncEvaluator(filterExpression);
+      // Evaluate the filter with the relevant quad binding
+      const result: boolean = await evaluator.evaluateAsEBV(bindings);
+      filterMap.set(relation.node, result);
+    }
+    return { filters: filterMap };
+  }
+
+  /**
+   *
+   * @param queryBody
+   * @param path
+   * @returns RDF.Quad[]
+   * find the quad that has as predicate a the TREE:path of a relation
+   */
+  private findRelavantQuad(queryBody: RDF.Quad[], path: string): RDF.Quad[] {
+    const resp: RDF.Quad[] = [];
+    for (const quad of queryBody) {
+      if (quad.predicate.value === path && quad.object.termType === 'Variable') {
+        resp.push(quad);
+      }
+    }
+    return resp;
+  }
+
+  /**
+   * @param filterExpression
+   * @param binding
+   * @returns Algebra.Expression
+   * delete the filters that are not related to TREE relation
+   */
+  private deleteUnrelevantFilter(filterExpression: Algebra.Expression, binding: Bindings): Algebra.Expression {
+    let newFilterExpression: Algebra.Expression = {
```

Comment:
Could you create this using sparqlalgebrajs's Factory?

Reply:
I think it's more complicated since I'm working with the inner expression.